### PR TITLE
Fix Alluxio inferring partitions for BooleanType with Hive

### PR DESCRIPTION
**Pull request by tgravescs**
fixes https://github.com/NVIDIA/spark-rapids/issues/6029

So when setting alluxio up, our code tries to replace paths specified with alluxio paths. The code to do this runs through creating a whole new FileIndex. This including inferring partitioning.  The Spark code we use to do this doesn't properly handle Boolean type (https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L433).    

To work around this I have changed the code to specifically handle certain file index types (CatalogFileIndex for hive and PartitionAware for regular file types with partitioning) and to only update the paths within those and not call the inferring code.  This handle this specific issue and the normal paths one would take for hive and reading files directly.   Unfortunately if we end up not knowing the file index type (like custom CSP file indexes) we use the old inferring code so it possible to still hit a case like this.   We have issue https://github.com/NVIDIA/spark-rapids/issues/6070 to followup.

Tested this manually on Databricks using different file types - hive, parquet directly, delta and all worked.  Also tested locally with reproduce case listed in the issue.

### CODE REVIEW

**Comment by tgravescs** [comment]
@GaryShen2008 @wjxiz1992

**Comment by tgravescs** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala [hunk 1]
```diff
@@ -289,48 +289,90 @@ object AlluxioUtils extends Logging {
...
-          replaceMapOption.get.values.find(value => rootPath.toString.startsWith(value)).foreach {
-            matched => checkAlluxioMounted(relation.sparkSession, matched)
+      // If we know the type of file index, try to reuse as much of the existing
+      // FileIndex as we can to keep from having to recalculate it and potentially
+      // mess it up. Things like partitioning are already included and some of the
+      // Spark infer code didn't handle certain types properly. Here we try to just
+      // update the paths to the new Alluxio path. If its not a type of file index
+      // we know then fall back to inferring. The latter happens on certain CSPs
+      // like Databricks where they have customer file index types.
+      if (relation.location.isInstanceOf[PartitioningAwareFileIndex]) {
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:320]: it seems we can replace this with a match statement, then `pfi` and `cfi` are obtained from the `case`.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala [hunk 2]
```diff
@@ -289,48 +289,90 @@ object AlluxioUtils extends Logging {
...
+        PartitionSpec(spec.partitionColumns, partitionsWithPathsReplaced)
+      }
 
-      // replace all of rootPaths which are already unique
-      val rootPaths = relation.location.rootPaths.map(replaceFunc.get)
+      def createNewFileIndexWithPathsReplaced(
+          spec: PartitionSpec,
+          rootPaths: Seq[Path]): InMemoryFileIndex = {
+        val specAdjusted = replacePathsInPartitionSpec(spec)
+        val replacedPaths = rootPaths.map(p => replaceFunc.get(p))
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:304]:
```suggestion
        val replacedPaths = rootPaths.map(replaceFunc.get)
```

I kind of wish we would `.get` the function at the top and store it in `replaceFunc` then change the original variable to `maybeReplaceFunc`.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala [hunk 3]
```diff
@@ -289,48 +289,90 @@ object AlluxioUtils extends Logging {
...
+        createNewFileIndexWithPathsReplaced(pfi.partitionSpec(), pfi.rootPaths)
+      } else if (relation.location.isInstanceOf[CatalogFileIndex]) {
+        logDebug("Handling CatalogFileIndex")
+        val cfi = relation.location.asInstanceOf[CatalogFileIndex]
+        val memFI = cfi.filterPartitions(Nil)
+        createNewFileIndexWithPathsReplaced(memFI.partitionSpec(), memFI.rootPaths)
+      } else {
+        logDebug(s"Handling file index type: ${relation.location.getClass}")
+        // With the base Spark FileIndex type we don't know how to modify it to
+        // just replace the paths so we have to try to recompute.
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:332]: is there a finite list of FileIndex types that we would handle here? E.g. could we refer to them so we know why we are not special casing them.

**Review by abellina — COMMENTED** [review]
A couple of minor comments.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala (hunk 3 above)
Comment (tgravescs) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:332]: I think the answer is yes... any datasource could create their own fileIndex type... like for delta for instance. If its not a type we know then we don't know what api they might be using to take into account partitioning and things like that.

**Comment by tgravescs** [comment]
build

**Comment by pxLi** [comment]
https://github.com/NVIDIA/spark-rapids/issues/6146 will try re-trigger

**Comment by pxLi** [comment]
build

**Comment by tgravescs** [comment]
look like issue connecting to databricks, rekicking

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
its failing to connect to databricks 10.4, I've manually tested building it there and running manually so going to remove databricks from title.

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
orc test failure, have issue for it

**Comment by tgravescs** [comment]
build

**Review by wjxiz1992 — APPROVED** [review]

**Pull request merged**